Fail test if a falsy value is thrown by the test

rankida opened the pull request:
This fixes #812 _awaiting a rejected promise can still pass test_. All the info can be found here.

Marsup commented on lib/runner.js:
```diff
@@ -578,7 +578,7 @@ internals.protect = function (item, state) {
             }
             catch (ex) {
-                return finish(ex);
+                return finish(ex || 'reject');
```
Shouldn't it be a `new Error()` ?

rankida replied:
If it is an `Error` then it will flow the full way through and report an empty string as the error, whereas setting it as a string results in the standard `Non Error object received or caught` message.

I am happy enough with either (my priority is that the test fail) so really depends on what feedback makes most sense

**with 'reject'**
```
Failed tests:

  27) Executor jobs throw:

      Non Error object received or caught

      at finish...
```

**with `new Error()`**
```
   27) Executor jobs throw:



      at Immediate.setImmediate...
```

geek replied:
This seems reasonable to me, @Marsup any objections to the current implementation?

Marsup replied:
OK, didn't investigate enough to tell, I was just surprised that it could end with an error or a string. If you're happy I'm happy.

lerouxb replied:
What's wrong with `new Error('Empty promise rejection.')`?

rankida replied:
As I said above, the "problem" with using an genuine `Error` is that you follow the same path as any errors that come from the test.

So you could see the error message `'Empty promise rejection.'` and think your test was actually throwing that.

The reason I chose a non-`Error` value is that I think the existing message `Non Error object received or caught` is the most appropriate and fits in nicely with the existing behaviour.

That said, as long as it is set to _something_ I do not care too much. My priority is that it fails the test.

rankida replied:
@geek if you want me to change anything just let me know

ORESoftware commented:
maybe dis better?

```js
return finish(ex || 'dummy hapijs reject');
```

rankida commented:
@ORESoftware thanks for your comment.

The string is used only internally to `lab` so I don't think it matters too much. If it does need changed I would prefer something that does not mention `hapi` as you might think this si something coming from the webserver.

I have addressed the comments about using `Error` in the thread above
